File: password strength checker and generator.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <time.h>

// Function to check password strength
void check_pwd(char *password) {
    int strength = 0;
    int lower_count = 0, upper_count = 0, num_count = 0, wspace_count = 0, special_count = 0;
    char remarks[5][50] = {""};

    // Iterate through each character in the password
    for (int i = 0; i < strlen(password); i++) {
        if (islower(password[i])) {
            lower_count++;
        } else if (isupper(password[i])) {
            upper_count++;
        } else if (isdigit(password[i])) {
            num_count++;
        } else if (isspace(password[i])) {
            wspace_count++;
        } else {
            special_count++;
        }
    }

    // Calculate password strength
    if (lower_count >= 1) strength++;
    if (upper_count >= 1) strength++;
    if (num_count >= 1) strength++;
    if (wspace_count >= 1) strength++;
    if (special_count >= 1) strength++;

    // Check for common patterns
    if (strstr(password, "abc") || strstr(password, "123") || strstr(password, "qwerty")) {
        strcpy(remarks[0], "Password contains common pattern.");
    }

    // Determine password strength remarks
    if (strength == 1) {
        strcpy(remarks[1], "Very Bad Password!!! Change ASAP");
    } else if (strength == 2) {
        strcpy(remarks[1], "Not A Good Password!!! Change ASAP");
    } else if (strength == 3) {
        strcpy(remarks[1], "It's a weak password, consider changing");
    } else if (strength == 4) {
        strcpy(remarks[1], "It's a hard password, but can be better");
    } else if (strength == 5) {
        strcpy(remarks[1], "A very strong password");
    }

    // Print password analysis
    printf("Your password has: \n");
    printf("%d lowercase characters\n", lower_count);
    printf("%d uppercase characters\n", upper_count);
    printf("%d numeric characters\n", num_count);
    printf("%d whitespace characters\n", wspace_count);
    printf("%d special characters\n", special_count);
    printf("Password Strength: %d\n", strength);
    for (int i = 0; i < 5; i++) {
        if (strlen(remarks[i]) > 0) {
            printf("Hint: %s\n", remarks[i]);
        }
    }
}

// Function to generate a strong password
void generate_strong_pwd(int length) {
    char password[length + 1];
    char alphabet[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()_+";

    srand(time(0));
    while (1) {
        for (int i = 0; i < length; i++) {
            password[i] = alphabet[rand() % (sizeof(alphabet) - 1)];
        }
        password[length] = '\0';

        int has_lower = 0, has_upper = 0, has_digit = 0, has_special = 0;
        for (int i = 0; i < length; i++) {
            if (islower(password[i])) has_lower = 1;
            else if (isupper(password[i])) has_upper = 1;
            else if (isdigit(password[i])) has_digit = 1;
            else has_special = 1;
        }
        if (has_lower && has_upper && has_digit && has_special) {
            break;
        }
    }
    printf("Generated strong password: %s\n", password);
}

// Function to ask user if they want to check another password
int ask_pwd(int another_pwd) {
    char choice[2];
    int valid = 0;

    if (another_pwd) {
        printf("Do you want to enter another pwd (y/n): ");
    } else {
        printf("Do you want to check pwd (y), generate strong pwd (g) or quit (q): ");
    }

    while (!valid) {
        scanf(" %1s", choice);
        if (strcmp(choice, "y") == 0) {
            return 1;
        } else if (strcmp(choice, "n") == 0) {
            return 0;
        } else if (strcmp(choice, "g") == 0) {
            generate_strong_pwd(12);
            return 0;
        } else if (strcmp(choice, "q") == 0) {
            return 0;
        } else {
            printf("Invalid, Try Again\n");
            if (another_pwd) {
                printf("Do you want to enter another pwd (y/n): ");
            } else {
                printf("Do you want to check pwd (y), generate strong pwd (g) or quit (q): ");
            }
        }
    }
    return 0;
}

// Main program
int main() {
    printf("+++ welcome to PWD checker +++\n");

    // Ask user if they want to check a password
    int ask_pw = ask_pwd(0);

    // Check password strength
    while (ask_pw) {
        char password[100];
        printf("Enter Password: ");
        scanf("%99s", password);
        check_pwd(password);
        ask_pw = ask_pwd(1);
    }

    return 0;
}

File: password strength checker and generator.py
import string
import secrets
import re
import tkinter as tk
from tkinter import messagebox

# Function to check password strength
def check_pwd(password):
    strength = 0
    remarks = []
    lower_count = upper_count = num_count = wspace_count = special_count = 0

    for char in password:
        if char in string.ascii_lowercase:
            lower_count += 1
        elif char in string.ascii_uppercase:
            upper_count += 1
        elif char in string.digits:
            num_count += 1
        elif char == ' ':
            wspace_count += 1
        else:
            special_count += 1

    if lower_count >= 1: strength += 1
    if upper_count >= 1: strength += 1
    if num_count >= 1: strength += 1
    if wspace_count >= 1: strength += 1
    if special_count >= 1: strength += 1

    if re.search(r"abc|123|qwerty", password, re.IGNORECASE):
        remarks.append("⚠️ Common pattern detected.")

    if strength == 1:
        remarks.append("❌ Very Bad Password! Change ASAP.")
    elif strength == 2:
        remarks.append("⚠️ Not a Good Password! Change ASAP.")
    elif strength == 3:
        remarks.append("🟡 Weak Password. Consider changing.")
    elif strength == 4:
        remarks.append("🟠 Decent Password. Could be stronger.")
    elif strength == 5:
        remarks.append("✅ Very Strong Password!")

    result = (
        f"Lowercase: {lower_count}\n"
        f"Uppercase: {upper_count}\n"
        f"Digits: {num_count}\n"
        f"Whitespace: {wspace_count}\n"
        f"Special: {special_count}\n"
        f"Strength Score: {strength}/5\n"
        + "\n".join(remarks)
    )
    return result

# Function to generate a strong password
def generate_strong_pwd(length=12):
    alphabet = string.ascii_letters + string.digits + string.punctuation
    while True:
        password = ''.join(secrets.choice(alphabet) for _ in range(length))
        if (any(c.islower() for c in password)
                and any(c.isupper() for c in password)
                and any(c.isdigit() for c in password)
                and any(c in string.punctuation for c in password)):
            return password

# GUI setup
def analyze_password():
    pwd = entry.get()
    if not pwd:
        messagebox.showwarning("Input Error", "Please enter a password.")
        return
    result = check_pwd(pwd)
    output.config(state='normal')
    output.delete(1.0, tk.END)
    output.insert(tk.END, result)
    output.config(state='disabled')

def generate_password():
    pwd = generate_strong_pwd()
    entry.delete(0, tk.END)
    entry.insert(0, pwd)
    output.config(state='normal')
    output.delete(1.0, tk.END)
    output.insert(tk.END, "Generated Password:\n" + pwd)
    output.config(state='disabled')

# Main window
root = tk.Tk()
root.title("🔐 Password Checker & Generator")
root.geometry("400x400")
root.resizable(False, False)

# Widgets
tk.Label(root, text="Enter Password:", font=("Arial", 12)).pack(pady=10)
entry = tk.Entry(root, show="*", width=30, font=("Arial", 12))
entry.pack()

tk.Button(root, text="Check Strength", command=analyze_password, bg="#4CAF50", fg="white").pack(pady=10)
tk.Button(root, text="Generate Strong Password", command=generate_password, bg="#2196F3", fg="white").pack(pady=5)

output = tk.Text(root, height=10, width=45, state='disabled', wrap='word', font=("Arial", 10))
output.pack(pady=10)

# Run the GUI
root.mainloop()
